### PR TITLE
Bernoulli Consistency Loss: physics coupling for p + 0.5|u|² = C

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The model predicts pressure (p) and velocity (Ux, Uy) as separate outputs with no explicit coupling between them. But for incompressible steady flow, these are coupled by the Bernoulli equation along a streamline:

```
p + 0.5 * (Ux² + Uy²) = p_total  (constant along streamline)
```

At the airfoil surface, the no-penetration condition reduces this to a relationship between wall pressure and tangential velocity. The current architecture uses a pressure-first sequential decoder (`--pressure_first --pressure_deep`) that conditions velocity on pressure predictions — but this coupling only exists in the forward pass, not in the loss. Adding a soft Bernoulli consistency auxiliary loss on surface nodes would make the p→u coupling explicit at the gradient level, providing physically-grounded signal that couples the two heads.

This is distinct from the failed Normal-Velocity Hard Constraint (PR #2187, hard-enforce u_n=0). A **soft** auxiliary loss is far less likely to destabilize training.

**Expected gains:** -1% to -3% p_in and surface velocity. The coupling may also improve p_tan if TE velocity prediction becomes more consistent with TE pressure.

## Instructions

Add a Bernoulli consistency auxiliary loss on surface nodes. The loss penalizes deviation of the model's own (p, Ux, Uy) predictions from the Bernoulli equation — no ground-truth labels needed.

### Step 1: Implement the Bernoulli loss

The pressure is predicted in asinh-transformed space (`--asinh_pressure --asinh_scale 0.75`). Invert the transform to get physical pressure before applying Bernoulli. Velocity predictions are in linear (un-transformed) space.

```python
def bernoulli_consistency_loss(pred_uvp, is_surface, Umag, asinh_scale, weight):
    """
    pred_uvp: [B, N, 3] — predicted (Ux, Uy, p) in model space
    is_surface: [B, N] bool — surface node mask
    Umag: [B] — freestream velocity magnitude per sample
    asinh_scale: float — the asinh transform scale factor
    weight: float — loss weight
    """
    if not is_surface.any():
        return torch.tensor(0.0, device=pred_uvp.device)

    ux = pred_uvp[..., 0]  # [B, N] — linear space
    uy = pred_uvp[..., 1]  # [B, N] — linear space
    p_transformed = pred_uvp[..., 2]  # [B, N] — asinh space

    # Invert asinh transform to physical pressure
    p_phys = torch.sinh(p_transformed / asinh_scale) * asinh_scale

    # Stagnation pressure: p_total = 0.5 * Umag^2 (inlet dynamic pressure, normalized)
    p_total = 0.5 * (Umag ** 2).unsqueeze(-1)  # [B, 1]

    # Bernoulli residual on surface nodes only
    q = 0.5 * (ux ** 2 + uy ** 2)
    bern_residual = (p_phys + q - p_total)[is_surface]  # should be ~0
    return weight * (bern_residual ** 2).mean()
```

### Step 2: Add CLI flags

```python
parser.add_argument('--bernoulli_loss', action='store_true')
parser.add_argument('--bernoulli_weight', type=float, default=0.01)
```

### Step 3: Call in training loop

After `pred_uvp` is computed in `train_step`, add:

```python
if cfg.bernoulli_loss:
    bern_l = bernoulli_consistency_loss(
        pred_uvp, is_surface, batch_Umag, cfg.asinh_scale, cfg.bernoulli_weight
    )
    loss = loss + bern_l
    # Log: metrics['train/bernoulli_loss'] = bern_l.item()
```

**Check how Umag is available per-sample.** It may be stored in the batch dict (e.g. `batch['Umag']`, `batch['vel_mag']`) or derivable from the boundary conditions. If not directly available, approximate from the inlet velocity features or from the AoA + freestream scalar. Inspect the batch structure in `train.py` to find it — it must exist since the model uses freestream conditions.

Ensure `pred_uvp` is the **non-EMA** model's output (not the EMA snapshot), so gradients flow through the Bernoulli loss.

### Step 4: Run experiments

```bash
# Primary run — bernoulli_weight=0.01, seed 42
cd cfd_tandemfoil && python train.py \
  --agent thorfinn --wandb_name "thorfinn/bernoulli-w0.01-s42" \
  --wandb_group bernoulli-consistency-loss \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --bernoulli_loss --bernoulli_weight 0.01 \
  --seed 42

# Run seed 73 with identical flags (change seed and wandb_name)
```

If seed 42 shows improvement, run weight=0.001 and weight=0.05 under the same `--wandb_group` to assess sensitivity. The `train/bernoulli_loss` should stay below ~10% of the main surface loss — if it dominates, reduce weight.

### Monitoring in W&B

Log `train/bernoulli_loss` every step. Key checks:
- Is the Bernoulli loss decreasing over training? (Model becoming more physically consistent)
- Does high Bernoulli loss correlate with high surface MAE samples?
- At convergence, what is the typical residual magnitude?

## Baseline

Current best (PR #2213, Wake Deficit Feature, 2-seed avg):

| Metric | Value | Merge threshold |
|--------|-------|-----------------|
| **p_in** | **11.979** | < 11.98 |
| p_oodc | 7.643 | < 7.65 |
| **p_tan** | **28.341** | < 28.34 |
| **p_re** | **6.300** | < 6.30 |

W&B: `hgml7i2r` (seed 42), `qic03vrg` (seed 73)

Reproduce baseline:
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-wake-deficit" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT thorfinn:

## Results

### Seed 42 (146 epochs, 180.9 min) — bernoulli_weight=0.01

| Metric | Baseline (2-seed avg) | Bernoulli s42 | Delta |
|--------|----------------------|---------------|-------|
| **p_in** | 11.979 | **23.274** | **+94.3% ↑ (catastrophic)** |
| **p_tan** | 28.341 | **33.4** | **+17.8% ↑ (worse)** |
| **p_oodc** | 7.643 | **14.0** | **+83.2% ↑ (catastrophic)** |
| **p_re** | 6.300 | **10.8** | **+71.4% ↑ (catastrophic)** |

**Full metrics (best at epoch 71/146):**
- val_in_dist: loss=0.4606, mae_surf_Ux=16.905, mae_surf_Uy=2.275, mae_surf_p=23.274
- val_tandem_transfer: loss=1.0247, mae_surf_p=33.4
- val_ood_cond: loss=0.3527, mae_surf_p=14.0
- val_ood_re: loss=0.3134, mae_surf_p=10.8

**Seed 73: Not run** — catastrophic regression, no point in second seed.
**Weight sweep: Not run** — failure mode is fundamental, not weight-dependent.

**W&B:** run ID `9guzym40`
**Peak VRAM:** 46.6 GB (no overhead from Bernoulli — it's a scalar loss)
**Command:**
```bash
cd cfd_tandemfoil && python train.py --agent thorfinn --wandb_name "thorfinn/bernoulli-w0.01-s42" \
  --wandb_group "bernoulli-consistency-loss" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --bernoulli_loss --bernoulli_weight 0.01 \
  --seed 42
```

## Analysis

**Catastrophic failure.** The Bernoulli consistency loss destroyed training, with surface pressure MAE nearly doubling (+94%). The model hit its best at epoch 71, then diverged severely — by epoch 146, val_ood_re had risen from 0.31 to 1.00 (3x worse than the best checkpoint).

**Why it failed — three compounding factors:**

1. **Bernoulli is wrong at viscous walls.** At the airfoil surface, the no-slip condition means Ux=Uy≈0, so the Bernoulli equation reduces to Cp≈1.0 everywhere. But in viscous flow (Re=1-5M), Cp varies significantly across the surface (suction peak, trailing edge pressure recovery, etc.). The loss was pushing the model toward physically incorrect predictions: constant Cp=1.0 at the wall with zero velocity. This explains the catastrophic surface velocity error (mae_surf_Ux=16.9 vs baseline ~3.2).

2. **Asymmetric gradient flow through pressure-first architecture.** The model uses `--pressure_first` with detached pressure conditioning velocity. The Bernoulli loss gradient could update the velocity head but NOT the pressure head (blocked by detach). This created an asymmetric training dynamic: velocity was being pulled toward Bernoulli consistency with a pressure prediction it couldn't modify, leading to degraded velocity predictions.

3. **Late-training instability.** The Bernoulli loss scale grew as predictions improved (larger velocity magnitudes → larger squared terms → larger gradients). This created a destabilizing feedback loop in the later epochs, visible as wild oscillations in val_ood_cond (0.35 → 1.00 between consecutive epochs).

**Key observation from researcher agent:** The literature warns of a "consistency barrier" (arxiv 2602.10611) when enforcing an approximate physics constraint that the true solution does not satisfy. RANS solutions at Re=1-5M with boundary layers violate inviscid Bernoulli at the surface, creating an irreducible error floor from the constraint itself.

## Suggested follow-ups
- **Variance form instead of anchored target:** minimize variance of `Cp + |u|²` across surface nodes rather than penalizing deviation from 1.0 — this couples the heads without asserting a physically incorrect target
- **Anneal weight to 0** in the last 20% of training to avoid late-stage instability
- **Volume nodes only** (away from boundary layer) where Bernoulli actually holds — surface nodes are the worst place to apply this
- **Much smaller weight** (0.001) with annealing could give a mild coupling effect without dominating

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2224 because Catastrophic failure — all metrics regressed 70-94%. Only seed 42 was run (rightly stopped). The Bernoulli equation is physically incorrect at viscous walls (no-slip → Cp≈1 everywhere, which contradicts the actual viscous Cp distribution). The pressure-first architecture also blocks gradient flow from the Bernoulli loss to the pressure head (detached), creating asymmetric training dynamics. Late-training instability from growing gradient magnitudes further compounds the damage.

Excellent diagnosis of the three failure modes. The 'consistency barrier' insight (Bernoulli doesn't hold where we're applying it) is the core lesson — physics losses must enforce constraints that the true solution actually satisfies. The student's suggested variance-form alternative is interesting conceptually but Bernoulli-based losses at the surface are fundamentally misguided for viscous flows.

Physics-informed loss direction at the surface is exhausted for this architecture/flow regime..
